src/runtime_src/core/pcie/linux/shim.cpp

```diff
@@ -211,53 +438,96 @@ class queue_cb {
             }
         }
 
+        if (rc < min_compl && qAioBatchEn) {
+             /* timeout happened, check if there is any io submission errors */
+             rc = check_io_submission_error(max_compl - num_evt, comps + num_evt);
+             num_evt += rc;
+        }
+
+        *actual = num_evt;
         return 0;
     }
 
     // submit the read/write i/o to the queue
     ssize_t queue_submit_io(xclQueueRequest *wr, aio_context_t *mAioCtx)
```

Comment:
This function is too long.  I can't quickly comprehend it.  Maybe refactor by creating synchronous and asynchronous submit_io functions ?

Reply:
I moved queuing up of the wr to a separate function.
In this function, it first checks if batching is enabled or not.  If it does and this i/o is async, it just queue up the new i/o. If this i/o is synchronous then first flush all of the queued async i/o to the kernel before submitting this sync i/o.
If no batching, just go ahead and submit to the kernel.